packages/itwin/tree-widget/src/components/trees/models-tree/internal/FilteredTree.ts

```diff
@@ -0,0 +1,255 @@
+    if (parentNodes.length === 0) {
+      return filterTargets;
+    }
+    lookupParents = parentNodes;
+  }
+
+  // find filtered nodes that match the `node`
+  const filteredNodes = findMatchingFilteredNodes(lookupParents, nodeKey.instanceKeys);
+  if (filteredNodes.length === 0) {
+    return filterTargets;
+  }
+
+  filteredNodes.forEach((filteredNode) => collectFilterTargets(filterTargets, filteredNode));
+  return filterTargets;
+}
+
+function findMatchingFilteredNodes(lookupParents: Array<{ children?: Map<Id64String, FilteredTreeNode> }>, keys: InstanceKey[]) {
+  return lookupParents
+    .flatMap((lookup) => keys.map((key) => lookup.children?.get(key.id)))
+    .filter((lookupNode): lookupNode is FilteredTreeNode => lookupNode !== undefined);
+}
+
+function collectFilterTargets(filterTargets: FilterTargets, node: FilteredTreeNode) {
+  if (node.isFilterTarget) {
+    addTarget(filterTargets, node);
+    return;
+  }
+
+  if (node.type === "element") {
+    // need to add parent ids as filter target will be an element
+    addTarget(filterTargets, node);
+  }
+
+  if (!node.children) {
+    return;
+  }
+
+  for (const child of node.children.values()) {
+    collectFilterTargets(filterTargets, child);
+  }
+}
+
+function addTarget(filterTargets: FilterTargets, node: FilteredTreeNode) {
+  switch (node.type) {
+    case "subject":
+      (filterTargets.subjects ??= new Set()).add(node.id);
+      return;
+    case "model":
+      (filterTargets.models ??= new Set()).add(node.id);
+      return;
+    case "category":
+      (filterTargets.categories ??= new Set()).add(createCategoryKey(node.modelId, node.id));
+      return;
+    case "element":
+      const categoryKey = createCategoryKey(node.modelId, node.categoryId);
+      const elements = (filterTargets.elements ??= new Map()).get(categoryKey);
+      if (elements) {
+        elements.add(node.id);
+        return;
+      }
+      filterTargets.elements.set(categoryKey, new Set([node.id]));
+      return;
+  }
+}
+
+function createFilteredTreeNode({
+  type,
+  id,
+  isFilterTarget,
+  parent,
+}: {
+  type: FilteredTreeNode["type"];
+  id: string;
+  isFilterTarget: boolean;
+  parent: FilteredTreeNode | FilteredTreeRootNode;
+}): FilteredTreeNode {
+  if (type === "subject" || type === "model") {
+    return {
+      id,
+      isFilterTarget,
+      type,
+    };
+  }
+
+  if (type === "category") {
+    assert("type" in parent && parent.type === "model");
+    return {
+      id,
+      isFilterTarget,
+      type,
+      modelId: parent.id,
+    };
+  }
+
+  if ("type" in parent && parent.type === "category") {
+    return {
+      id,
+      isFilterTarget,
+      type,
+      modelId: parent.modelId,
+      categoryId: parent.id,
+    };
+  }
+
+  if ("type" in parent && parent.type === "element") {
+    return {
+      id,
+      isFilterTarget,
+      type,
+      modelId: parent.modelId,
+      categoryId: parent.categoryId,
```

Comment:
Looks like this is already existing issue. Filled a new issue to avoid increasing scope of this PR: https://github.com/iTwin/viewer-components-react/issues/1100